### DISCUSSION TITLE
Planned obsoletion of Application.UrlsOpened with IActivatableLifetime and file type association on desktop OSes

### DISCUSSION THREAD

**Discussion by xecrets** [AvaloniaUI/Avalonia#15272] [Ideas]
Hi!

Starting with 11.0.8 (I think) the `Application.UrlsOpened` event handling file type association based launching of an app is marked as obsolete, and the plan seems to partially replace it with more extensive support for protocol-based activation such as supported by macOS, iOS and Android - but not Linux and Windows. That's nice as far as that goes of course!

However, in desktop OSes Linux/Windows/macOS, the concept of file type association is supported in a way similar for users - double clicking a "document" in whatever desktop shell is used, results in the "document" being "opened" in the associated default application. Variations exist where you can pick and chose between other existing associations, and other mechanisms for invoking the association often exist as well.

On Windows and Linux, this is typically implemented by providing the application "command line" with arguments, such as one or more files to "open". On macOS it works very differently technically, but similar from a user viewpoint.

Currently,  this is neatly wrapped in the `Application.UrlsOpened` event, which hides the details of the underlying implementation from the application, and makes it easy to implement cross platform support for file type association and "opening" of documents on all the desktop OSes.

The plan now seems to be to abandon this approach, and instead implement application activation in a more general way - but only on platforms where this concept exists, i.e. macOS, iOS and Android.

Since desktop file type association and protocol activation are two different scenarios it's not necessarily an improvement for all to remove the support for `Application.UrlsOpened` (file type association) and referring to support for `IActivatableLifetime` for a different scenario (protocol activation) instead.

I'd like to suggest that support for `Application.UrlsOpened` is retained, or in some other way supported, so that desktop applications can support file type association without having to resort to private implementations and run time platform specific code.

**Comment by kekekeks** [comment]
@maxkatz6

**Comment by maxkatz6** [comment]
We previously discussed this topic in the telegram chat.
Just created an issue https://github.com/AvaloniaUI/Avalonia/issues/15316 to discuss this API extensions. 
I am already working on FileActivatedEventArgs to be added in 11.1. But addition of CommandLineActivatedEventArgs needs to be discussed.
